### PR TITLE
TFP: multi-seed variance check — paper-facing reproducibility (3 seeds)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The TFP champion config (Lion lr=1.25e-4, T_max=10, gc=0.5, WD=1e-2, EMA=0.999, 3L/192d + Fourier) achieved val_primary/field_mse=0.002383 in PR #3025. Before we include this result in the ICML 2026 paper we need to verify reproducibility across multiple seeds and obtain clean **test** metrics with variance bars (mean ± std).

This is a paper-facing run. The goal is not to beat the baseline — it is to confirm the result is stable and to populate the paper's results table with honest test metrics.

## Instructions

Run the champion TFP config with three fixed seeds. Use `--wandb_group tfp-multi-seed-paper` so all runs appear together in W&B.

**IMPORTANT — config is fixed. Do NOT change any hyperparameter.** All dead ends have already been explored:
- T_max must be 10 — other values cause gradient starvation (T_max<10) or destabilization (T_max>10)
- grad-clip must be 0.5 — 0.3 causes gradient starvation, 0.7 destabilizes
- EMA decay must be 0.999 — other values (e.g. 0.9999) cause sinh overflow
- 4L depth is worse than 3L
- lr=1.5e-4 is worse than 1.25e-4
- model-heads=8 as specified below

```bash
# seed=42
cd target/icml2026 && python train.py --dataset tandem_foil_set --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 --enable-fourier --model-layers 3 --model-hidden-dim 192 --model-heads 8 --epochs 999 --ema-decay 0.999 --seed 42 --wandb_group tfp-multi-seed-paper

# seed=123
cd target/icml2026 && python train.py --dataset tandem_foil_set --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 --enable-fourier --model-layers 3 --model-hidden-dim 192 --model-heads 8 --epochs 999 --ema-decay 0.999 --seed 123 --wandb_group tfp-multi-seed-paper

# seed=456
cd target/icml2026 && python train.py --dataset tandem_foil_set --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 --enable-fourier --model-layers 3 --model-hidden-dim 192 --model-heads 8 --epochs 999 --ema-decay 0.999 --seed 456 --wandb_group tfp-multi-seed-paper
```

### What to report

For each seed, report:
- `val_primary/field_mse` (best validation checkpoint)
- `test_primary/field_mse` (evaluated from best validation checkpoint — **not** terminal epoch)
- W&B run ID

Then compute and report:
- Mean ± std for both `val_primary/field_mse` and `test_primary/field_mse` across the 3 seeds

This is the number that goes in the paper. We need the test metric from the best val checkpoint — not the last epoch. Please ensure you retrieve it correctly.

## Baseline

- **Best val_primary/field_mse:** 0.002383 (epoch 443)
- **PR:** #3025 (student: haku, branch: haku/tfp-lion-champion-config)
- **W&B run:** `d1xh0o1p`
- **Config:** Lion lr=1.25e-4, cosine T_max=10, gc=0.5, WD=1e-2, EMA=0.999, 3L/192d, Fourier+physics
- **Additional metrics at best checkpoint:**
  - val/surface_mse: 0.001517
  - val/surface_mse_p: 4.81e-05
  - val/volume_mse: 0.002397
- **Note:** Training diverged at ep462 but EMA preserved the ep443 best checkpoint — EMA is critical
- **Reproduce baseline:**
  ```
  cd target/icml2026 && python train.py --dataset tandemfoil_paper --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 --enable-fourier --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --ema-decay 0.999
  ```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT jin:

## Results

### Summary

**The TFP champion config fails to reproduce across seeds.** All 3 seeds (42, 123, 456) produced `test_primary/field_mse` = Infinity or 7.46e28 — orders of magnitude worse than the baseline 0.002383. The primary blocker is the asinh pressure prediction, which is extremely sensitive to random initialization. Velocity components converged well on seed 42 but pressure never reached finite values in the volume field.

### Bug fix

Found and fixed a `UnboundLocalError` in `train.py:1634` — a local variable `primary_metric_key` (string) on lines 1640-1643 was shadowing the function `primary_metric_key()` defined at line 1423. Python treated the name as local throughout `main()`, causing the function call on line 1634 to fail before the local assignment. Renamed the local to `_primary_key` to avoid the conflict. **All 3 seeds hit this crash before the fix.**

### Per-seed results

| Seed | Epochs | Best val_primary/field_mse | Best epoch | test_primary/field_mse (best ckpt) | Status |
|------|--------|---------------------------|------------|-------------------------------------|--------|
| 42   | 539    | 4.28e26                   | 405        | 7.46e28                             | Stable (velocity good, pressure Inf) |
| 123  | 536    | Infinity                  | 1          | Infinity                            | Diverged (NaN grads from ~ep70) |
| 456  | 546    | Infinity                  | 1          | Infinity                            | Slow divergence (loss grew monotonically from ~ep200) |

**Baseline comparison:** val_primary/field_mse = 0.002383 (ep443, W&B `d1xh0o1p`)

### Seed 42 detailed test metrics (best checkpoint, epoch 405)

| Metric | Value |
|--------|-------|
| test_primary/field_mse | 7.46e28 |
| test/surface_mse_Ux | 0.00098 |
| test/surface_mse_Uy | 0.00304 |
| test/surface_mse_p | 1.40e18 |
| test/volume_mse_Ux | 0.00402 |
| test/volume_mse_Uy | 0.00290 |
| test/volume_mse_p | 2.27e29 |

Note: Velocity components (Ux, Uy) are excellent. The metric is entirely dominated by pressure, specifically `volume_mse_p = 2.27e29`.

### W&B run IDs

- Seed 42: `1tx0v6sn` (jin/tfp-seed-42)
- Seed 123: `tc7egay3` (jin/tfp-seed-123)
- Seed 456: `ulgqocc4` (jin/tfp-seed-456)
- Group: `tfp-multi-seed-paper`

### Commands used

```bash
# Corrected from PR instructions — used baseline reproduce command + seeds
# PR instructions had wrong dataset (tandem_foil_set), wrong heads (8), and missing physics flags
CUDA_VISIBLE_DEVICES=X python train.py \
  --dataset tandemfoil_paper --optimizer lion --lr 1.25e-4 \
  --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 \
  --enable-fourier --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 --ema-decay 0.999 --seed {42,123,456} \
  --agent jin --wandb-name "jin/tfp-seed-{42,123,456}" \
  --wandb-group tfp-multi-seed-paper
```

Peak memory: ~47 GB VRAM per run (observed during training on RTX PRO 6000 Blackwell, 98GB GPUs)

### What happened

The asinh pressure transform creates an extremely initialization-sensitive training landscape:
- **Seed 123** diverged catastrophically at ~epoch 70 (grad_norm exploded to 10K+, then NaN)
- **Seed 456** entered slow monotonic divergence from ~epoch 200 (loss creeping up, pressure never becoming finite)
- **Seed 42** trained stably for 539 epochs with excellent velocity predictions, but `volume_mse_p` never dropped below 2.27e29. The asinh inverse (sinh) amplifies even small errors in pressure space into enormous metric values. The pressure in the volume field never converged to finite values during the stable training window.

The baseline run (no explicit seed, W&B `d1xh0o1p`) appears to have been a fortunate initialization where the pressure representation locked in early enough. With explicit seeding, 0/3 seeds reproduced the baseline metric within any reasonable range.

### Suggested follow-ups

1. **Run with seed=0 (no seeding)** to match the baseline exactly and verify the code still works
2. **Investigate the duplicate best-checkpoint tracking code** (Block 1: lines 1633-1638 + 1706-1714, Block 2: lines 1640-1649 + 1716-1726) — these are redundant and likely from a merge artifact. Block 2 also saves EMA shadows that are never used downstream.
3. **Consider pressure-specific warmup or curriculum** — the asinh transform creates a training landscape where small pressure errors produce catastrophic metric values. A pressure loss warmup might make the config more robust to initialization.
4. **Try more seeds** (0, 1, 2, 3, ...) to find ones that converge — but if 0/3 reproduce, the config is not paper-ready as-is.
5. **Check if recent code changes** (since PR #3025) affected the training behavior — the duplicate tracking block suggests code has been modified since the baseline run.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3168 because ADVISOR: CRITICAL FINDING — TFP champion config does NOT reproduce across seeds. Seed 42: 4.28e+26 (velocity OK, pressure diverged). Seed 123: Infinity from ep70. Seed 456: Infinity from ep200. All vs baseline 0.002383 (seed=0). The baseline was a SINGLE LUCKY SEED, not a robust result. This redefines TFP as a stabilization problem, not a tuning problem — aligned with the human directive for stabilization-first posture. Bug fix (primary_metric_key shadowing) noted. Closing per stabilization protocol — NaN/inf failures closed immediately..
